### ISSUE TITLE
[proposal] pattern for handling subgraph and node data

### ISSUE THREAD

**Issue by cnasc** [with-backed/backed-interface#281]
By example, how I think we should handle data fetching

## Library code
- `lib/loans/node/nodeLoanById.ts`
  - This fetches all the loan info from the node, then normalizes it into a common compatible `Loan` instance
- `lib/loans/subgraph/subgraphLoanById.ts`
  - This fetches the loan info from the subgraph, and likewise normalizes it
  - In the event we get a null result from the subgraph, we fall back to `nodeLoanById` to see if it's something the Graph just hasn't indexed yet.

## API routes
- `pages/api/loans/[id].ts`
  - this route will only ever be called from the client (e.g., while refreshing loan info) so it _only talks to the node_ by using `nodeLoanById`.

## Server-side rendering
`getServerSideProps` calls `subgraphLoanById` directly and returns an object like
```javascript
const subgraphLoan = await subgraphLoanById(7);
return {
  props: {
    fallback: {
      '/api/loans/7': subgraphLoan,
    },
  },
}
```

## Client-side use
We will have a `fetcher` function that wraps `fetch` and calls `parseLoanInfoJson` on the result.

```jsx
const fetcher = ...;

function Loan() {
  // `data` will always be available as it's in `fallback`.
  // we can force a data refresh by calling `mutate` with no arguments
  const { data, mutate } = useSWR('/api/loans/7', fetcher)
  return <h1>{data.id}</h1>
}

export default function Page({ fallback }) {
  // SWR hooks inside the `SWRConfig` boundary will use those values.
  return (
    <SWRConfig value={{ fallback }}>
      <Loan />
    </SWRConfig>
  )
}
```

## tl;dr
- API routes _do not_ query the graph
- `getServerSideProps` prefers the graph, but will fall back to the node
- client-side we use [SWR](https://swr.vercel.app/docs/with-nextjs) to handle the SSR fallback and refetching
